src/controllers/video.controller.js

```diff
@@ -1,47 +1,321 @@
+    ){
+        throw new ApiError(400, "update fields are required")
+    }
+
+    // find video 
+    const previousVideo = await Video.findOne(
+        {
+            _id: videoId
+        }
+    )
+    if(!previousVideo){
+        throw new ApiError(404, "video not found")
+    }
+
+    let updateFields = {
+        $set:{
+            title,
+            description,
+        }
+    }
+
+    // if thumbnail provided delete the previous one and upload new on 
+    let thumbnailUploadOnCloudinary;
+    if(thumbnailFile){
+        await deleteOnCloudinary(previousVideo.thumbnail?.public_id)
+
+        // upload new one 
+         thumbnailUploadOnCloudinary = await uploadOnCloudinaary(thumbnailFile);
+
+        if(!thumbnailUploadOnCloudinary){
+            throw new ApiError(500, "something went wrong while updating thumbnail on cloudinary !!")
+        }
+
+        updateFields.$set = {
+            public_id: thumbnailUploadOnCloudinary.public_id,
+            url: thumbnailUploadOnCloudinary.url
+        }
+    }
+
+    const updatedVideoDetails = await Video.findByIdAndUpdate(
+        videoId,
+        updateFields,
+        {
+            new: true
+        }
+    )
+
+    if(!updatedVideoDetails){
+        throw new ApiError(500, "something went wrong while updating video details");
+    }
+
+    //retrun responce
+    return res.status(200).json(new ApiResponse(
+        200,
+        { updatedVideoDetails },
+        "Video details updated successfully!"
+    ));
+
 })
 
+// get video by id
 const getVideoById = asyncHandler(async (req, res) => {
     const { videoId } = req.params
-    //TODO: get video by id
+
+    if(!isValidObjectId(videoId)){
+        throw new ApiError(400, "This video id is not valid")
+    } 
+
+    const video = await Video.findById(
+        {
+            _id: videoId
+        }
+    )
+
+    if(!video){
+        throw new ApiError(404, "video not found")
+    }
+
+    // return responce
+    return res.status(200).json(
+        new ApiResponse(
+            200,
+            video,
+            "video fetched successfully!!"
+        )
+    )
 })
 
-const updateVideo = asyncHandler(async (req, res) => {
-    const { videoId } = req.params
-    //TODO: update video details like title, description, thumbnail
+// get all video 
+const getAllVideos = asyncHandler(async (req, res) => {
+    const { 
+
+        page = 1,
+        limit = 10,
+        query = `/^video/`,
+        sortBy = "createdAt",
+        sortType = 1, 
+        userId = req.user._id } = req.query
+
+    // find user in db
+    const user = await User.findById(
+        {
+            _id: userId
+        }
+    )
+
+    if(!user){
+        throw new ApiError(404, "user not found")
+    }
+
+    const getAllVideosAggregate = await Video.aggregate([
+        {
+            $match: { 
+                videoOwner: new mongoose.Types.ObjectId(userId),
```

Comment:
owner: new mongoose.Types.ObjectId(userId)

the schema of video is 
import mongoose, { Schema } from "mongoose";
import mongooseAggregatePaginate from "mongoose-aggregate-paginate-v2";

const videoSchema = new Schema(
  {
    videoFile: {
      type: String, //cloudinary url
      required: true,
    },
    thumbnail: {
      type: String, //cloudinary url
      required: true,
    },
    title: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    duration: {
      type: Number,
      required: true,
    },
    views: {
      type: Number,
      default: 0,
    },
    isPublished: {
      type: Boolean,
      default: true,
    },
    owner: {
      type: Schema.Types.ObjectId,
      ref: "User",
    },
  },
  {
    timestamps: true,
  }
);

videoSchema.plugin(mongooseAggregatePaginate);

export const Video = mongoose.model("Video", videoSchema);